### PR TITLE
Feature/grpc plugin

### PR DESCRIPTION
**Pull request by colindickson**

### CODE REVIEW

**Review by maoueh — CHANGES_REQUESTED** [review]
We should use `buf` to generate the Go bindings. `substreams` project is using `buf` in the `generate` command, you can port it over. I have probably other examples here and there.

File: proto/sf/metering/v1/metering.proto [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  rpc Emit(Event) returns (google.protobuf.Empty) {}
+}
+
+message Event {
+  string user_id = 1;
+  string api_key_id  = 2;
+  string ip_address = 3;
+
+  // Defines the service that emitted the event (sf.firehose.v1/Blocks ...)
+  string endpoint = 4;
```
Comment (maoueh) [inline proto/sf/metering/v1/metering.proto]: I know you did not choose the name ourself, but I'm wondering if this should not be called `service`, at least the documentation seems to say so ...

File: types.go [hunk 2]
```diff
@@ -1,16 +1,114 @@
...
+	"net/url"
+
+	pbmetering "github.com/streamingfast/dmetering/pb/sf/metering/v1"
+
+	"github.com/streamingfast/dgrpc"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+)
+
+func init() {
```
Comment (maoueh) [inline types.go]: We should really refrain from using `init`. I much prefer to expose a `RegisterGRPC` from the interface and do the `func init() { dmetering.RegisterGRPC() }` in the calling side (e.g. `firehose-<chain>`) than implicit `init`.

File: types.go [hunk 3]
```diff
@@ -1,16 +1,114 @@
...
+		logger:  logger,
+	}, nil
+}
+
+func (g *grpcEmitter) Emit(ctx context.Context, ev Event) error {
+	pbevent := ev.ToProtoMeteringEvent()
+	pbevent.Network = g.network
+
+	if pbevent.Network == "" || pbevent.Endpoint == "" {
+		g.logger.Warn("events must contain network and endpoint. dropping event", zap.Object("event", ev))
```
Comment (maoueh) [inline types.go]: Dropping event but still emitting it.

Comment (maoueh) [inline types.go]:
```suggestion
		g.logger.Warn("events must contain network and endpoint, dropping event", zap.Object("event", ev))
```

File: types.go [hunk 4]
```diff
@@ -1,16 +1,114 @@
...
+	_, err := g.client.Emit(ctx, pbevent)
+	if err != nil {
+		return fmt.Errorf("unable to emit event: %w", err)
+	}
+	return nil
+}
+
+func newMeteringClient(endpoint string) (pbmetering.MeteringClient, error) {
+	var dialOptions []grpc.DialOption
+	conn, err := dgrpc.NewExternalClient(endpoint, dialOptions...)
```
Comment (maoueh) [inline types.go:58]: `conn` close handling is missing, we should have a way to gracefully terminate the client connection.

File: zlog_plugin.go [hunk 5]
```diff
@@ -1,89 +0,0 @@
-package dmetering
-
-import (
```
Comment (maoueh) [inline zlog_plugin.go:3]: Don't we want to keep this one, it was quite useful for testing purposes. I'm not against the delete, but it seems we could easily keep it.

**Review by maoueh — CHANGES_REQUESTED** [review]
More review

File: zlog_plugin.go (hunk 5 above)
Comment (colindickson) [reply zlog_plugin.go:3]: this functionality is now in the `loggerEmitter`

File: buf.gen.yaml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+version: v1
+managed:
+  enabled: false
+plugins:
+  - name: go
```
Comment (maoueh) [inline buf.gen.yaml:5]: We should use explicit remote plugin so the version is controlled

File: pb/generate.sh [hunk 7]
```diff
@@ -1,74 +1,197 @@
...
+    echo ""
+    echo "If everything is working as expetcted, the command:"
+    echo ""
+    echo "  buf --version"
+    echo ""
+    echo "Should print '1.11.0' (or newer)"
+    exit 1
+  fi
+
   # The old `protoc-gen-go` did not accept any flags. Just using `protoc-gen-go --version` in this
```
Comment (maoueh) [inline pb/generate.sh:146]: Using buf remote plugins removes the need for all those checks as we would rely only on `buf`.

File: pb/generate.sh [hunk 8]
```diff
@@ -1,74 +1,197 @@
-#!/bin/bash
+##!/bin/bash
+#
+#ROOT="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && cd .. && pwd )"
```
Comment (maoueh) [inline pb/generate.sh:3]: Let's remove the commented out part before merging PR.

File: types.go (hunk 2 above)
Comment (maoueh) [reply types.go]: What about this one? Maybe we discuss together later.

**Review by maoueh — CHANGES_REQUESTED** [review]

File: zlog_plugin.go (hunk 5 above)
Comment (maoueh) [reply zlog_plugin.go:3]: I see, I missed it indeed. I preferred to have them in their own file, it make it easily visible from the project structure what are the valid implementation of an Emitter. But you moved them, so maybe you preferred having all them in a file. I leave you the decision here.

**Pull request merged**